File: frontend/src/api/audit_logs.ts
import { apiFetch } from './client'
import type { AuditLogRead } from '../types'

export interface AuditLogFilters {
  action?: string
  resource_type?: string
  status?: string
  since?: string
  until?: string
  limit?: number
}

export function getAuditLogs(filters: AuditLogFilters = {}): Promise<AuditLogRead[]> {
  const params = new URLSearchParams()
  if (filters.action) params.set('action', filters.action)
  if (filters.resource_type) params.set('resource_type', filters.resource_type)
  if (filters.status) params.set('status', filters.status)
  if (filters.since) params.set('since', filters.since)
  if (filters.until) params.set('until', filters.until)
  if (filters.limit != null) params.set('limit', String(filters.limit))
  const qs = params.toString()
  return apiFetch<AuditLogRead[]>(`/api/audit-logs${qs ? `?${qs}` : ''}`)
}

File: backend/app/api/audit_logs.py
import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import require_admin
from app.models.audit_log import AuditLog
from app.models.user import User
from app.schemas.audit_log import AuditLogRead

router = APIRouter(prefix="/audit-logs", tags=["audit-logs"])


@router.get("", response_model=list[AuditLogRead])
async def list_audit_logs(
    user_id: uuid.UUID | None = Query(default=None),
    action: str | None = Query(default=None),
    resource_type: str | None = Query(default=None),
    status: str | None = Query(default=None),
    since: datetime | None = Query(default=None),
    until: datetime | None = Query(default=None),
    limit: int = Query(default=100, le=500),
    session: AsyncSession = Depends(get_db),
    _admin: User = Depends(require_admin),
) -> list[AuditLogRead]:
    stmt = select(AuditLog).order_by(AuditLog.created_at.desc()).limit(limit)
    if user_id is not None:
        stmt = stmt.where(AuditLog.user_id == user_id)
    if action is not None:
        stmt = stmt.where(AuditLog.action == action)
    if resource_type is not None:
        stmt = stmt.where(AuditLog.resource_type == resource_type)
    if status is not None:
        stmt = stmt.where(AuditLog.status == status)
    if since is not None:
        stmt = stmt.where(AuditLog.created_at >= since)
    if until is not None:
        stmt = stmt.where(AuditLog.created_at <= until)
    result = await session.execute(stmt)
    return [AuditLogRead.model_validate(row) for row in result.scalars().all()]
